lib/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+use crate::core::File;
+
+/// The supported TEE types:
+/// - TDX: TDX TEE.
+/// - SGX: SGX TEE.
+/// - SEVSNP: SEV-SNP TEE.
+/// - SAMPLE: A dummy TEE that used to test/demo the attestation service functionalities.
+#[derive(Debug)]
+pub enum TEE {
+    TDX,
+    SGX,
+    SEVSNP,
+    SAMPLE,
+}
+
+#[derive(Debug)]
+pub struct Service {
+    pub attestation: core::Attestation,
+}
+
+impl Service {
+    /// Create a new attestation service's instance.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use attestation_service::Service;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let service = Service::new();
+    ///     assert!(service.is_ok());
+    /// }
+    /// ```
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            attestation: core::Attestation::default(),
+        })
+    }
+
+    /// Attest the received Evidence by the attestation service instance.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use attestation_service::Service;
+    /// use serde_json::json;
+    /// use sha2::{Sha384, Digest};
+    /// use base64;
+    ///
+    /// // sample TEE's evidence
+    /// fn evidence() -> String {
+    ///     let nonce = "the nonce".to_string();
+    ///     let public_key = "the public key".to_string();
+    ///     let pubkey = json!({
+    ///         "algorithm": "".to_string(),
+    ///         "pubkey-length": "".to_string(),
+    ///         "pubkey": public_key
+    ///     }).to_string();
+    ///     let mut hasher = Sha384::new();
+    ///     hasher.update(&nonce);
+    ///     hasher.update(&pubkey);
+    ///     let hash = hasher.finalize();
+    ///     let tee_evidence = json!({
+    ///         "is_debuggable": false,
+    ///         "cpusvn": 1,
+    ///         "svn": 1,
+    ///         "report_data": base64::encode(hash)
+    ///     }).to_string();
+    ///     json!({
+    ///         "nonce": nonce,
+    ///         "tee": "sample".to_string(),
+    ///         "tee-pubkey": pubkey,
+    ///         "tee-evidence": tee_evidence
+    ///     }).to_string()
+    /// }
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let service = Service::new().unwrap();
+    ///
+    ///     // Attest the evidence with default OPA policy and reference data.
+    ///     let res = service.attestation(&evidence(), None, None).await;
+    ///     assert!(res.is_ok());
+    ///
+    ///     let policy = r#"
+    /// package policy
+    /// default allow = false
+    ///
+    /// allow {
+    ///     input.cpusvn >= data.cpusvn
+    ///     input.svn >= data.svn
+    /// }
+    /// "#.to_string();
+    ///
+    ///     let reference_data = json!({
+    ///        "cpusvn": 1,
+    ///        "svn": 1
+    ///     }).to_string();
+    ///
+    ///     // Attest the evidence with customized OPA policy and reference data.
+    ///     let res = service.attestation(&evidence(), Some(policy), Some(reference_data)).await;
```

Comment:
That looks much better now. We get an evaluation based on an evidence, a policy and some reference data.

Reply:
Yes, the `attesation()` can receivce policy and reference data file dynamically now. If the policy/reference data file is None, it will use the default file.